src/afa/afa.cc

```diff
@@ -1003,41 +1002,127 @@ void Mata::Afa::construct(
 
 	// do the dishes and take out garbage
 	clean_up();
-} // construct }}}
 
+    return aut;
+} // construct }}}
 
-void Mata::Afa::construct(
-	Afa*                                 aut,
-	const Mata::Parser::ParsedSection&  parsec,
-	StringToSymbolMap*                   symbol_map,
-	StringToStateMap*                    state_map)
+Afa Mata::Afa::construct(
+        const Mata::IntermediateAut&         inter_aut,
+        Alphabet*                            alphabet,
+        StringToStateMap*                    state_map)
 { // {{{
-	assert(nullptr != aut);
+    Afa aut;
+    assert(nullptr != alphabet);
 
-	bool remove_symbol_map = false;
-	if (nullptr == symbol_map)
-	{
-		symbol_map = new StringToSymbolMap();
-		remove_symbol_map = true;
-	}
+    if (!inter_aut.is_afa()) {
+        throw std::runtime_error(std::string(__FUNCTION__) + ": expecting type \"" +
+                                 Mata::Afa::TYPE_AFA + "\"");
+    }
 
-	auto release_res = [&](){ if (remove_symbol_map) delete symbol_map; };
+    bool remove_state_map = false;
+    if (nullptr == state_map) {
+        state_map = new StringToStateMap();
+        remove_state_map = true;
+    }
 
-  Mata::Nfa::OnTheFlyAlphabet alphabet(*symbol_map);
+    // a lambda for translating state names to identifiers
+    auto get_state_name = [&state_map, &aut](const std::string& str) {
+        if (!state_map->count(str)) {
+            State state = aut.add_new_state();
+            state_map->insert({str, state});
+            return state;
+        } else {
+            return (*state_map)[str];
+        }
+    };
+
+    // a lambda for cleanup
+    auto clean_up = [&]() {
+        if (remove_state_map) { delete state_map; }
+    };
+
+    // lambda returning true if node is operator of given type
+    auto is_node_operator =
+            [](const FormulaNode& node, FormulaNode::OperatorType type) -> bool {
+        return node.is_operator() && node.operator_type == type;
+    };
+
+    // lambda creates Node from set of strings which are state names
+    auto create_node = [get_state_name](
```

Comment:
I myself have no issue with this approach (i use it in python quite often), though, I'm wondering whether Mr. L would be satisfied with this and would prefer normal functions (e.g., global).

Reply:
This is just a filter for programmer intelligence.

Reply:
I'm afraid certain Mr. L. will fail it then :D

Reply:
that's life